### PR TITLE
Added support to use decorators in Typescript typed Jasmine Tests

### LINKED ISSUE #59: Report Steps with Annotation/Decorator on TypeScript PageObjects
As a user I want scenarioo to automatically report a step on each call to functions on page objects, with a description of that function as step text, if that  function is annotated with a typescript decorator `reportStep`, such that I do not have to write the description of the step manually anymore.

Accpetance criteria
* calls to page object functions annotated with decorator `reportStep` are automatically reported as steps into scenarioo documentation
* all usual step parameters can be passed on the steps decorator.
* by default the description of the step is generated from the page object class name followed by the name of the page object function

### PR DESCRIPTION
**Pull request by sebastianzillessen**
Added support to use decorators in Typescript typed Jasmine Tests to perform step annotation automatically after a method of a page object gets executed.

Fixes #59

### CODE REVIEW

File: example/test/pages/startPage.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  clickFirstListItem() {
+    return this.firstListItem.click();
+  }
+
+  /**
+   * By default, the @stepAnnotation takes the Class Name followed by the method name as a description
+   * (e.g. 'StartPage: clickSecondListItem').
+   * This behaviour can be overwritten by providing a custom description as follows.
+   */
+  @stepAnnotation("Custom message for this step") // this will result in the description: 'Custom message for this step'
```
Comment (bruderol) [inline example/test/pages/startPage.ts]: Could we rename the annotation please to just "step" ? I think "stepAnnotation" is unnecessary long, I see that it is an annotation like that.

File: README.md [hunk 2]
```diff
@@ -184,6 +185,7 @@ Or you can try to do this by hooking into protractor functions, to ensure that a
 
 We recommend to do it in the page objects, because that is usually the place where you know, that something worthy of recording as a step happened.
 
+If you are using Typescript (old Jasmine syntax or the new Fluent DSL), you can follow the example [exampleBasicJasmineWithTypeScript.spec.ts](example/test/exampleBasicJasmineWithTypeScript.spec.ts) and the used page object [startPage.ts](example/test/pages/startPage.ts) as a reference on how to use a [Typescript Decorator](http://www.typescriptlang.org/docs/handbook/decorators.html) to perform this in an easy manner.
```
Comment (bruderol) [inline README.md]: I would prefer to integrate that example into our full Fluent DSL example of how we recommend to write tests, such that we have that example in there as well.

File: lib/index.d.ts [hunk 3]
```diff
@@ -45,6 +45,27 @@ export declare var scenario: (name: string) => ScenarioDeclaration;
...
+/**
+ * Saves a step in your e2e tests.
+ *
+ * Use this decorator (http://www.typescriptlang.org/docs/handbook/decorators.html) this in your e2e test functions
+ * whenever you want scenarioo to report a step based on a call of a function on a page object.
+ *
+ * @param {string} [stepDescription?] - optional description text for the step to be recorded, will be displayed in `title` field of a step in scenarioo.
+ * if not provided, it will use the following pattern: `{objectName}: {methodName}`.
+ */
+export declare var stepAnnotation: (stepDescription?: string) => any;
```
Comment (bruderol) [inline lib/index.d.ts]: would be cool if also additionalProperties could be pased, where necessary, example if I want to put labels on a step,as in the usual step function

File: example/test/pages/startPage.ts (hunk 1 above)
Comment (bruderol) [reply example/test/pages/startPage.ts]: Oh - I see "step" is allready used --- hmmm, how about "scStep" or "reportStep" - any better idea?

**Comment by sebastianzillessen** [comment]
Hei @bruderol, thanks for your feedback. I implemented the suggested changes!

**Comment by sebastianzillessen** [comment]
@bruderol any updates on this?

**Comment by bruderol** [comment]
@sebastianzillesen sorry, we enjoyed the summer and have been busy otherwise ;-) sorry for letting you wait. I will try to get it merged and released soon.

File: example/test/exampleFluentDslWithTypeScriptAndPageObjects.spec.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      .description('An optional but recommended description for the scenario')
+      .labels(['happy', 'example-label'])
+      .it(() => {
+        // the StartPage constructor navigates us to index.html
+
+        // A step on a decorator could also have additional propoerties, like e.g. labels
+        startPage.assertHeaderShown();
+
+        // do something on the page itself.
+        startPage.clickFirstListItem();
```
Comment (danielsuter) [inline example/test/exampleFluentDslWithTypeScriptAndPageObjects.spec.ts]:
```suggestion
        startPage.selectFirstListItem();
```
We should write test on use case or scenario level. If the goal is to select the element, we should name it accordingly. It does not matter how we do it (e.g. click).

**Comment by danielsuter** [comment]
@bruderol imho we can merge this pr. I only have some small comments.

**Comment by sebastianzillessen** [comment]
Thanks @danielsuter for your comments. I applied them and pushed the latest version.

**Comment by danielsuter** [comment]
@sebastianzillessen Only one thing is remaining. After I merged the greenkeeper branch, please verify if the demo tests still behave as expected. At the moment we can't do this automatically. I will notify you, when it's done.
You can't verify this at the moment as they are broken.

**Comment by danielsuter** [comment]
@sebastianzillessen You missed nothing. We have tests in the example directory. Those are not checked automatically by the CI as some of them are expected to fail. 
These tests are not working at the moment but will as soon as my PR is merged (Friday). You work is absolutely fine but I can't merge this one before the other.

**Comment by danielsuter** [comment]
@sebastianzillessen 
Can you please rebase/merge the changes in your branch and check, if you get the following output in the example tests:
https://github.com/scenarioo/scenarioo-js/tree/develop/example

14 specs, 5 failures, 5 pending specs

**Comment by sebastianzillessen** [comment]
Hi @danielsuter .
I did solve the merge conflicts and did run the tests in `/example`. The output is as follows:

> 16 specs, 5 failures, 5 pending specs

I would say this is expected, as I added 2 more tests. 
The readme is upgraded regarding this new tests.

Ready to merge? ☺️

**Comment by danielsuter** [comment]
@sebastianzillessen Thank you very much for your effort.

**Pull request merged**